File: app/src/main/java/eu/toma/dev/playground/warcraft/mounts/MountCache.java
package eu.toma.dev.playground.warcraft.mounts;

import java.util.List;

import javax.inject.Inject;

import eu.toma.dev.playground.warcraft.di.MountScope;
import io.reactivex.Observable;

@MountScope
public class MountCache
{
    private List<Mount> cachedMounts;

    @Inject
    public MountCache()
    {
    }

    public Observable<List<Mount>> getCachedMounts()
    {
        if(cachedMounts != null){
            return Observable.fromIterable(cachedMounts).toList().toObservable();
        } else {
            return Observable.error(new Exception("no cached mounts available"));
        }
    }

    public void save(List<Mount> mounts)
    {
        this.cachedMounts = mounts;
    }
}

File: app/src/main/kotlin/warcraft/mounts/MountCache.kt
package warcraft.mounts

import javax.inject.Inject

import io.reactivex.Observable

@warcraft.mounts.di.Mounts
class MountCache

@Inject
constructor()
{
    private var cachedMounts: List<Mount>? = null

    fun getCachedMounts(): Observable<List<Mount>> {
        if (cachedMounts != null) {
            return Observable.fromIterable(cachedMounts).toList().toObservable()
        } else {
            return Observable.error<List<Mount>>(Exception("no cached mounts available"))
        }
    }

    fun save(mounts: List<Mount>) {
        this.cachedMounts = mounts
    }
}
